### PR TITLE
Bug 108780 submit evaluators if absent

### PR DESCRIPTION
**Pull request by gergopapi2**

### CODE REVIEW

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/context/EvaluatorContext.java [hunk 1]
```diff
@@ -3,4 +3,6 @@
 public interface EvaluatorContext {
 
     Object getData();
+
+    long getClusterOrStackId();
```
Comment (mhmxs) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/context/EvaluatorContext.java]: Maybe `getItemId` would be a better name

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/CronTimeEvaluator.java [hunk 2]
```diff
@@ -67,11 +78,15 @@ private boolean isPolicyAttached(BaseAlert alert) {
 
     @Override
     public void run() {
-        long start = System.currentTimeMillis();
-        Cluster cluster = clusterService.find(clusterId);
-        MDCBuilder.buildMdcContext(cluster);
-        publishIfNeeded(alertRepository.findAllByCluster(clusterId));
-        LOGGER.info("Finished cronTimeEvaluator for cluster {} in {} ms", clusterId, System.currentTimeMillis() - start);
+        try {
+            long start = System.currentTimeMillis();
```
Comment (mhmxs) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/CronTimeEvaluator.java]: Can move out of try, like AmbariAgentHealthEvaluator

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            LOGGER.info("submitting {} for cluster {} failed, it has been present in the system for {} ms",
+                    elementAlreadyPresent.getKey().getEvaluatorExecutor().getName(),
+                    elementAlreadyPresent.getKey().getClusterId(),
+                    clock.getCurrentTime() - elementAlreadyPresent.getTimestampMillis());
+            return false;
+        }
+    }
+
+    public void remove(EvaluatorExecutor evaluator, long clusterId) {
+        synchronized (REMOVE_MONITOR) {
```
Comment (mhmxs) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java]: Synchronized block is too wide

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    @Value(value = "${periscope.monitor.exeutor.registry.timeout:600000}")
+    private long timeout;
+
+    @Inject
+    private Clock clock;
+
+    private Map<Key, Element> submittedEvaluators = new ConcurrentHashMap<>();
+
+    public boolean putIfAbsent(EvaluatorExecutor evaluator, long clusterOrStackId) {
+        synchronized (PUT_MONITOR) {
```
Comment (mhmxs) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java]: Synchronized block is too wide

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
+package com.sequenceiq.periscope.monitor.executor;
```
Comment (mhmxs) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java]: Fix inspections please

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/context/EvaluatorContext.java (hunk 1 above)
Comment (gergopapi2) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/context/EvaluatorContext.java]: renamed to getItemId

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/CronTimeEvaluator.java (hunk 2 above)
Comment (gergopapi2) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/CronTimeEvaluator.java]: I moved it out of the try block

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java (hunk 3 above)
Comment (gergopapi2) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java]: removed synch completely

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java (hunk 4 above)
Comment (gergopapi2) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java]: removed synch block completely

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java (hunk 5 above)
Comment (gergopapi2) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/executor/EvaluatorExecutorRegistry.java]: fixed.

**Pull request merged**